src/rpc/context.rs

```diff
@@ -0,0 +1,309 @@
+};
+use serde::{Deserialize, Serialize};
+use std::{net::SocketAddr, ops::Deref, sync::Arc, time::Instant};
+use tokio::sync::{oneshot, RwLock};
+
+#[doc(hidden)]
+pub struct RpcInner<N: Network, E: Environment> {
+    pub(crate) address: Option<Address<N>>,
+    pub(crate) peers: Arc<Peers<N, E>>,
+    pub(crate) ledger: LedgerReader<N>,
+    pub(crate) operator: Arc<Operator<N, E>>,
+    pub(crate) prover_router: ProverRouter<N>,
+    pub(crate) memory_pool: Arc<RwLock<MemoryPool<N>>>,
+    /// RPC credentials for accessing guarded endpoints
+    #[allow(unused)]
+    pub(crate) credentials: RpcCredentials,
+    pub(crate) launched: Instant,
+}
+
+/// Implements RPC HTTP endpoint functions for a node.
+#[derive(Clone)]
+pub struct RpcContext<N: Network, E: Environment>(Arc<RpcInner<N, E>>);
+
+impl<N: Network, E: Environment> Deref for RpcContext<N, E> {
+    type Target = RpcInner<N, E>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<N: Network, E: Environment> RpcContext<N, E> {
+    /// Creates a new struct for calling public and private RPC endpoints.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        username: String,
+        password: String,
+        address: Option<Address<N>>,
+        peers: Arc<Peers<N, E>>,
+        ledger: LedgerReader<N>,
+        operator: Arc<Operator<N, E>>,
+        prover_router: ProverRouter<N>,
+        memory_pool: Arc<RwLock<MemoryPool<N>>>,
+    ) -> Self {
+        Self(Arc::new(RpcInner {
+            address,
+            peers,
+            ledger,
+            operator,
+            prover_router,
+            memory_pool,
+            credentials: RpcCredentials { username, password },
+            launched: Instant::now(),
+        }))
+    }
+}
+
+/// Defines the authentication format for accessing private endpoints on the RPC server.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct RpcCredentials {
+    /// The username in the credential
+    pub username: String,
+    /// The password in the credential
+    pub password: String,
+}
+
+/// RPC metadata for encoding authentication.
+#[derive(Default, Clone)]
+pub struct Meta {
+    /// An optional authentication string for protected RPC functions.
+    pub auth: Option<String>,
+}
+
+/// Starts a local RPC HTTP server at `rpc_port` in a dedicated `tokio` task.
+/// RPC failures do not affect the rest of the node.
+pub async fn initialize_rpc_server<N: Network, E: Environment>(
+    rpc_server_addr: SocketAddr,
+    rpc_server_context: RpcContext<N, E>,
+) -> (SocketAddr, tokio::task::JoinHandle<()>) {
+    let access_control = AccessControlBuilder::default().build(); // TODO(ljedrz): adjust to only accept the desired hosts?
+
+    let server = HttpServerBuilder::new()
+        .set_access_control(access_control)
+        .register_resource(ALL_CONCURRENT_REQUESTS, ALL_CONCURRENT_REQUESTS_LIMIT, 1) // Limit the number of requests handled at a time.
```

Comment:
Suggestion: comment on the fact that `1` here is the default?

Reply:
`1` is actually not the default here - the default would be `0`; I can add an extra comment though

Reply:
unless you meant that it's effectively the default value for every call, which it is; yeah, a comment will help :smile: